EuiDescribedFormGroup component

jen-huang opened the pull request:
`EuiDescribedFormGroup` is a wrapper around `EuiFormRow` that uses `EuiFlexGroup` to display help text horizontally next to the field instead of below (falls back to stacking on mobile). Includes support for validation errors and full width. If full width is not enabled, it maxes out at 800px (`$euiFormMaxWidth*2`).

Most of its props are passed to `EuiFormRow` with the following exceptions:
- `gutterSize` passed to `EuiFlexGroup`, defaults to `l`
- `helpTitle` used by itself, title that displays above the help text - it is treated as a second label for the field (or only label if `label` is ommited)
- `helpText` used by itself, not passed to form row

**Desktop view with focusing fields:**

![apr-26-2018 17-21-56](https://user-images.githubusercontent.com/1965714/39338223-6d5d7e1c-4976-11e8-9947-b2468bfe028d.gif)

**Mobile view:**

<img width="468" alt="screen shot 2018-04-26 at 5 24 11 pm" src="https://user-images.githubusercontent.com/1965714/39338272-b5ef583a-4976-11e8-8d17-1bb664a91b83.png">

chandlerprall commented:
If this were added to a form layout with non-descriptive elements the flow would be interrupted because of the descriptive row's different width and positioning of the input element. It might make sense to allow the two columns to be flipped in these cases so the input elements still align - thoughts @cchaos?

jen-huang commented:
@cchaos Go for it!

chandlerprall commented:
@cchaos Two more questions for stacked inputs:

1. How should the label/input IDs work in a stacked-input setup? As is with this PR, the descriptive label corresponds with the input in its row; is this desired?

2. How would the multiple elements be stacked under a single heading? I imagine using EuiDescriptiveFormRow and then separately creating a flex group as a new row, using the second column for the remaining inputs?

cjcenizal commented:
I love that we're adding this @jen-huang! I can use this to build a layout I had to hack together in Cloud:

![image](https://user-images.githubusercontent.com/1238659/39387487-aaa4b034-4a2e-11e8-85bc-5bc93a1e4a5b.png)

I have a few concerns though.

* @chandlerprall's point about mixing regular form rows with described form rows resonates with me. For example, what if I wanted the describe the "Default Value" control in the above screenshot, but not the "Available Increments" control? I don't think I could do that with this component. But being able to mix-and-match would give me that flexibility.
* I feel like we're conflating concerns a bit. The component is called `EuiDescriptiveFormRow` but it incorporates `fieldset` concepts using `role="group"`. I intuit rows to be _children_ of a group, not groups themselves.
* This component presents title-like text, but doesn't allow the consumer to specify a heading element to incorporate the title into the page's heading hierarchy. We could solve this by allowing the user to specify the heading via a prop, e.g. `titleEl="h3"`, but...

...I think it might make more sense to simplify this component, and make its focus solely on appending more help text to the form row. If it placed this text to the right then I think this would make this component play more nicely with regular form rows _and_ would make more sense in terms of IA since users should hopefully be able to just use the form, and only refer to the text if they're hopelessly confused or new to the form.

If we were to do this, then my heading hierarchy concern would be addressed because consumers would just use regular `EuiTitle` and `EuiText` to add their own titles. If we want to provide a `fieldset` solution (which I think would definitely be useful) then we can do that with another component, which would be focused on just that one concerns.

What do y'all think??

jen-huang commented:
@cjcenizal Some thought on your thoughts:

- I removed auto-associating with the latest update. The text of the component automatically gets assigned ID of `idFromProp-legend-text`, but it is not automatically associated with the input element. To associate it with the input, `EuiFormRow` needs `describedByIds={['idFromProp-legend-text']}`.
- `EuiFormRowGroup` might be a more meaningful name, if we want this component to wrap around arbitrary form groups.
- Keeping `title` as always `h3` lets us add correct amount of padding above the `EuiFormRow`s so that the top of the field is flush with the `text`. If we don't care about keeping these two aligned, I'm fine with removing `title` and deferring to user for build their own `text`. (We could also make `title` optional and only add padding when that prop is there, maybe?)

For reference, the use case for this component originally came up from https://github.com/elastic/kibana/pull/17465. Here's the design suggested by @cchaos as a way to show both human readable name of the setting as well as its code-like key:

<img width="1504" alt="screen shot 2018-04-16 at 13 05 25 pm" src="https://user-images.githubusercontent.com/549577/38824319-dc709efe-4176-11e8-92b8-646ddb47d042.png">

cjcenizal commented:
> I removed auto-associating with the latest update

If we adjust our approach to be that this component really is just a form row with help text appended to it (instead of thinking of this as a component involving multiple rows) then we can bring this functionality back, which I think would be a nice accessibility win.

We can uncouple the appearance of the title from the heading element it uses, similarly to how `EuiTitle` works. We'll need to allow the consumer to specify various heading elements to make sure the title fits within the heading hierarchy, since we shouldn't use `h3` unless an `h2` and `h1` already exist on the page, or an `h4` unless all three of those already exist, etc.

jen-huang commented:
That is basically the first iteration of this component, except not having title prop. (Which I'm fine with removing, and we can provide the user with a padding size prop to help with aligning to field.)

We changed the component to behave like fieldset to satisfy [this design pattern](https://user-images.githubusercontent.com/549577/39373405-7fb3e30e-4a15-11e8-8288-4d5659556047.png), which is super similar to [Cloud's use case](https://user-images.githubusercontent.com/1238659/39387487-aaa4b034-4a2e-11e8-85bc-5bc93a1e4a5b.png). The only difference is the description and input columns being flipped around.

@cchaos If we change EuiDescriptiveFormRow to have input on the left and description on the right, then we can avoid having to wrap multiple EuiFormRows for the inputs to be aligned. I guess even if we didn't want to flip them, the user can use EuiDescriptiveFormRow with an empty description to get them aligned... what do you think?

cjcenizal commented:
That's exactly what I was thinking too, Jen. I think the only difference between our cases is that my use case doesn't have descriptive text for a fieldset. If it were to have descriptive text for the fieldset, I think something like this would work:

![image](https://user-images.githubusercontent.com/1238659/39411733-1333c3dc-4bc5-11e8-94cb-f3927a6539ba.png)

This kind of layout will accommodate form layouts with fieldset descriptive text, field descriptive text, or neither.

As a side note, I've found the help text which we use with fields to generally be more useful for displaying validation rules, e.g. "Only letters and numbers allowed." I think of the descriptive text as the kind of stuff you would normally hide behind an [EuiIconTip](https://elastic.github.io/eui/#/display/tooltip) next to the label, but because the layout is so airy you can surface it directly in the layout. Maybe as a later step we could make bake in accessibility by automatically hiding it inside of EuiIconTips on narrower screens?

snide commented:
Agree with @cchaos on this one . Think we'll need to add some guidelines for forms fairly soon as well to solidify the best practices she's proposing.

cjcenizal commented:
jenkins test this

jen-huang commented:
Latest update has the following changes:

- Renamed `text` prop to `description`
- `title` and `description` are wrapped with `EuiText` and are both required
- `title` doesn't have any inherent styling or size
- Added `paddingSize` prop (`s`, `m`, or `l`) to help with aligning first field to `description`, user can set this to `null` if they don't want padding or want to specify it via their own rules

I've attempted to illustrate the relationship between `title` and `paddingSize` in the docs. Screenshot shows:

- `EuiDescriptiveFormRow` with `title={<h3/>}` and `paddingSize="m"`
- `EuiDescriptiveFormRow` with `title={<strong/>}` and `paddingSize="s"`
- `EuiDescriptiveFormRow` with `title={<h2/>}` and `paddingSize="l"`

![screen shot 2018-05-01 at 12 03 11 pm](https://user-images.githubusercontent.com/1965714/39488667-d0bdccae-4d37-11e8-85fa-291fb6cbe453.png)

@cchaos Welcoming any additional feedback or doc changes!

cjcenizal commented on src-docs/src/views/form_layouts/form_rows_descriptive.js:
```diff
@@ -0,0 +1,190 @@
+        >
+          <EuiFormRow
+            label="Text field"
+            describedByIds={['single-example-description']}
```
As a consumer, I'd have to know about the implementation details of EuiDescriptiveFormRow to know that `single-example-description` is an available ID, and that it's associated with the element containing the description. As a maintainer of the component, I may not be aware that consumers have created dependencies upon this ID and I might change it unwittingly and break screen-reader accessibility for a bunch of UIs (which we probably wouldn't discover for awhile).

I think the solution is to apply the ID directly to the EuiFlexItem component which encompasses the title and the description. This way the consumer can just reference the same ID they're passing in already. I tested this and it seems to work well.

If we make this change, we'd have to update the documentation too.

jen-huang replied:
I changed it so that the user will pass in `idAria` (React didn't like "ariaId") prop, which gets put on the `EuiText` description. They can pass the same ID to `describedByIds` so there won't be a case where they have to know how the description ID is generated.

If they pass `id` it'll just unfold to the top level `div` container for `EuiDescribedFormRow` 🙂

cjcenizal commented on src/components/form/form_row_descriptive/descriptive_form_row.test.js:
```diff
@@ -0,0 +1,145 @@
+      .toMatchSnapshot();
+
+    // Descriptive form row has group role and is labelled by its title and text.
+    expect(tree.find(`EuiFlexItem`).at(0).prop(`id`)).toEqual(`test-id-legend`);
```
I think we can remove lines 56-73 since they're all covered by the snapshot. If someone makes a change to the component in the future which affects any of the things we're asserting on these lines, then the snapshot comparison will also fail.

jen-huang replied:
removed

cchaos commented on src/components/form/form_row/form_row.js:
```diff
@@ -60,6 +59,7 @@ export class EuiFormRow extends Component {
       hasEmptyLabelSpace,
       fullWidth,
       className,
+      describedByIds,
```
That's fine. We can keep it in but we should have @aphelionz check it to make sure it doesn't duplicate the reading of the title and description.

cjcenizal replied:
Good call, I'd love to hear @aphelionz's thoughts on this. I tested it in VoiceOver on OS X and this does repeat the title and description, but I think that's just what happens when you associate elements with other elements using `aria-labelledby` and `aria-desrcribedby`. They'll be read once as a regular part of the page text and then again as information associated with a specific element.

aphelionz replied:
Can you add `aria-hidden` or some such thing to one of the pairs? It's not the worst thing in the world to read something twice (better than not reading it at all) but not ideal.

cjcenizal replied:
I'm not sure... the dilemma is that the title and description may or may not contain heading text. Here's a simplified representation of one scenario:

```html
<h3 id="description">This describes the input</h3>
<label for="input">Text input</label>
<input id="input" aria-describedby="description" type="text" />
```

In this scenario, we'd want to preserve the heading hierarchy of that `h3`, right? Would adding `aria-hidden` to it remove it from the hierarchy?

There are also scenarios where this is not a heading at all, in which case I think `aria-hidden` can be added without doing any harm:

```html
<p id="description">This describes the input</p>
<label for="input">Text input</label>
<input id="input" aria-describedby="description" type="text" />
```

cchaos replied:
Actually headings and descriptions are both required as those are integral to this component.

cjcenizal replied:
Sorry, to clarify the `<p>` in my second example is the result of the user specifying `title={<p>This describes the input</p>}`

aphelionz replied:
I've been staring at the WCAG docs trying to figure out which criterion this would fall under.... is the text _required_ to be understood succinctly by a user before requiring input? Then it's a level A... however if it's not it falls to level AA.

Level A: https://www.w3.org/WAI/WCAG20/quickref/#minimize-error-cues

vs

Level AA: https://www.w3.org/WAI/WCAG20/quickref/#qr-navigation-mechanisms-descriptive

Maybe even AAA:

cchaos replied:
Maybe we should ensure that users use heading tags for the title by doing:
```
  title: PropTypes.node.isRequired,
  headingLevel: PropTypes.oneOf(HEADING_LEVELS).isRequired,
```

cjcenizal replied:
I think @jen-huang's latest changes are sound and result in a good-enough UX. Maybe we can come back and reassess accessibility later on?

jen-huang replied:
I adjusted the aria and IDing a bit with the last update, but I think more consideration is needed for handling accessibility. I created #762 for more work 🙂

jen-huang commented on src/components/form/form_row_descriptive/descriptive_form_row.js:
```diff
@@ -0,0 +1,105 @@
+      >
+        <EuiFlexGroup gutterSize={gutterSize}>
+          <EuiFlexItem id={`${id}-legend`} grow={false}>
+            <EuiText size="xs" id={`${id}-title`} className="euiDescriptiveFormRow__title">
```
Updated, size will use prop `titleSize` but default to `xs` per Dave's suggestion

jen-huang commented on src/components/form/form_row_descriptive/descriptive_form_row.js:
```diff
@@ -0,0 +1,105 @@
+      >
+        <EuiFlexGroup gutterSize={gutterSize}>
+          <EuiFlexItem id={`${id}-legend`} grow={false}>
+            <EuiText size="xs" id={`${id}-title`} className="euiDescriptiveFormRow__title">
```
Added!

jen-huang commented on src-docs/src/views/form_layouts/form_layouts_example.js:
```diff
@@ -81,6 +87,26 @@ export const FormLayoutsExample = {
+    text: (
+      <p>
+        Use <EuiCode>DescriptiveFormRows</EuiCode> component to associate multiple <EuiCode>EuiFormRows</EuiCode>.
+        It can also simply be used with one <EuiCode>EuiFormRows</EuiCode> as a way to display help text (or additional
```
fixed

jen-huang commented on src-docs/src/views/form_layouts/form_rows_descriptive.js:
```diff
@@ -0,0 +1,190 @@
+  EuiFieldText,
+  EuiForm,
+  EuiFormRow,
+  EuiDescriptiveFormRow,
```
renamed to `EuiDescribedFormGroup` 😁

jen-huang commented on src/components/form/form_row_descriptive/descriptive_form_row.js:
```diff
@@ -0,0 +1,105 @@
+      >
+        <EuiFlexGroup gutterSize={gutterSize}>
+          <EuiFlexItem id={`${id}-legend`} grow={false}>
+            <EuiText size="xs" id={`${id}-title`} className="euiDescriptiveFormRow__title">
```
I changed the ID'ing a little, but now the title and description IDs are both referenced by the parent group in aria props.

jen-huang commented:
jenkins test this

cjcenizal commented:
CI failure looks like it's https://github.com/elastic/infra/issues/4813

jen-huang commented:
jenkins test this
